File: flask_app/app/services/product_service.py
from app.dao.product_dao import ProductDAO
from app.utils.discount_validation import check_product_discount_eligibility


class ProductService:

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        items, total = ProductDAO.get_all(
            page=page, per_page=per_page, filters=filters or {}
        )

        items_dicts = [item.to_dict() for item in items]

        for item in items_dicts:
            eligible, discount_value = check_product_discount_eligibility(item)
            # print(f'item: {item.get("name")} Discount Id: {item.get("discount_id")}  eligible: {eligible}, discount_value: {discount_value}')
            if eligible:
                item["discount_applied"] = True
                item["discount_value"] = discount_value
                item["discount_id"] = item.get("discount_id")
            else:
                item["discount_applied"] = False
                item["discount_value"] = 0
        return items_dicts, total

    @staticmethod
    def get_by_id(product_id):
        return ProductDAO.get_by_id(product_id)

    @staticmethod
    def create(data):
        return ProductDAO.create(data)

    @staticmethod
    def update(product_id, data):
        return ProductDAO.update(product_id, data)

    @staticmethod
    def commit(product):
        return ProductDAO.commit(product)

    @staticmethod
    def delete(product_id):
        return ProductDAO.soft_delete(product_id)

File: java_membership_api/src/main/java/com/example/membership_api/service/ProductService.java
package com.example.membership_api.service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.membership_api.dao.ProductDAO;
import com.example.membership_api.model.Product;

import jakarta.transaction.Transactional;

@Service
@Transactional
public class ProductService {

    @Autowired
    private ProductDAO productDAO;

    public Map<String, Object> getAll(int page, int perPage, Map<String, Object> filters) {
        List<Product> products = productDAO.getAll(page, perPage, filters);

        long total = productDAO.getTotalCount(filters);

        Map<String, Object> response = new HashMap<>();
        response.put("items", products);
        response.put("total", total);
        return response;
    }

    public Product getById(String productId) {
        return productDAO.getById(productId);
    }

    public Product create(Map<String, Object> data) {
        Product product = new Product();
        applyUpdates(product, data);
        return productDAO.create(product);
    }

    public Product update(String productId, Map<String, Object> data) {
        Product existing = productDAO.getById(productId);
        if (existing == null) return null;
        applyUpdates(existing, data);
        productDAO.commit(existing);
        return existing;
    }

    public void commit(Product product){
        productDAO.commit(product);
    }

    public boolean delete(String productId) {
        return productDAO.softDelete(productId);
    }

    private void applyUpdates(Product product, Map<String, Object> data) {
        if (data.containsKey("name")) product.setName((String) data.get("name"));
        if (data.containsKey("description")) product.setDescription((String) data.get("description"));
        if (data.containsKey("category_id")) product.setCategoryId((String) data.get("category_id"));
        if (data.containsKey("stock_quantity")) product.setStockQuantity((Integer) data.get("stock_quantity"));
        if (data.containsKey("discount_id")) product.setDiscountId((String) data.get("discount_id"));
        if (data.containsKey("price")) product.setPrice(Double.valueOf(data.get("price").toString()));
        if (data.containsKey("is_membership_exclusive"))
            product.setIsMembershipExclusive((Boolean) data.get("is_membership_exclusive"));
        if (data.containsKey("min_tier_required"))
            product.setMinTierRequired((Integer) data.get("min_tier_required"));
    }
}
